### PR TITLE
Fix detectOrthogonalDimRanges on DistributedMesh

### PR DESCRIPTION
**Pull request by roystgnr**
Previously this method would work reliably only while the orthogonal
mesh was still serialized, which was fine for most use cases, but
which could cause serious inconsistencies in corner cases.  I caught
this via a failure of bcs/periodic.auto_wrap_2d_test_non_generated
with --distributed-mesh --recover: the recovered mesh would have
incorrect bounds detected on some processors when enough processors
were employed, which would cause periodic BC offsets to be
miscalculated, which would cause periodic neighbor lookups to find
incorrect neighbor elements, which would fail libMesh consistency
checks when the neighbor elements lacked the correct periodic BC ID.

After this and the next libMesh update are merged, the state of #8410
should go from "passes all tests on up to 4 processors" to "passes all
tests on up to 6 processors".

### CODE REVIEW

**Comment by roystgnr** [comment]
Oh, and to be less confusing: *this* fix works for me up to 24 processors at least, it's just that there's still a failure in a different test at 6 processors.

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -1233,9 +1233,12 @@ MooseMesh::detectOrthogonalDimRanges(Real tol)
     }
   }
 
+  this->comm().max(max);
+  this->comm().min(min);
+
   _extreme_nodes.resize(8); // 2^LIBMESH_DIM
   // Now make sure that there are actual nodes at all of the extremes
-  unsigned int extreme_matches = 0;
+  std::vector<bool> extreme_matches(8, false);
```
Comment (permcody) [inline framework/src/mesh/MooseMesh.C]: Minor, but this is one of those cases where `std::array` could be used.

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -1260,12 +1263,13 @@ MooseMesh::detectOrthogonalDimRanges(Real tol)
     if (coord_match == dim) // Found a coordinate at one of the extremes
     {
       _extreme_nodes[comp_map[X] * 4 + comp_map[Y] * 2 + comp_map[Z]] = &node;
-      ++extreme_matches;
+      extreme_matches[comp_map[X] * 4 + comp_map[Y] * 2 + comp_map[Z]] = true;
```
Comment (permcody) [inline framework/src/mesh/MooseMesh.C]: I like this better.

File: framework/src/mesh/MooseMesh.C (hunk 1 above)
Comment (roystgnr) [inline framework/src/mesh/MooseMesh.C]: Hmm... actually it isn't, even though it should be: we don't currently have libMesh::Parallel::Communicator overloads for std::array.  I've finally got user unanimity on making libMesh C++11-only, but I haven't yet bit the bullet and broken C++98.

Comment (roystgnr) [inline framework/src/mesh/MooseMesh.C]: We could use vector<unsigned char> here, though, if you prefer.  IIRC @jwpeterson finds vector<bool> offensive (as do lots of C++ users, with reason), and this isn't a huge vector where the space savings matter.

Comment (permcody) [inline framework/src/mesh/MooseMesh.C]: No, this is fine, just saw the fixed size and had my knee-jerk reaction. I'm fine with vector<bool> in non-threaded regions.

**Review by permcody — APPROVED** [review]

**Pull request merged**